### PR TITLE
Change memory layout

### PR DESCRIPTION
**Pull request by matt-frey**

### CODE REVIEW

File: src/3d/parcels/parcel_diagnostics.f90 [hunk 1]
```diff
@@ -228,25 +228,25 @@ end subroutine calc_parcel_diagnostics
...
-!                     zbv = zbv + bv * parcels%position(n, 2)
+!                     xbv = xbv + bv * parcels%position(1, n)
+!                     zbv = zbv + bv * parcels%position(2, n)
 !
-!                     x2bv = x2bv + bv * parcels%position(n, 1) ** 2
-!                     z2bv = z2bv + bv * parcels%position(n, 2) ** 2
-!                     xzbv = xzbv + bv * parcels%position(n, 2) * parcels%position(n, 2)
+!                     x2bv = x2bv + bv * parcels%position(1, n) ** 2
+!                     z2bv = z2bv + bv * parcels%position(2, n) ** 2
+!                     xzbv = xzbv + bv * parcels%position(2, n) * parcels%position(2, n)
```
Comment (sjboeing) [inline src/3d/parcels/parcel_diagnostics.f90:239]: Is this actually right? May need to be: 
xzbv + bv * parcels%position(1, n) * parcels%position(2, n)

File: src/3d/parcels/parcel_diagnostics.f90 [hunk 2]
```diff
@@ -228,25 +228,25 @@ end subroutine calc_parcel_diagnostics
...
-!                     zvv = zvv + vv * parcels%position(n, 2)
+!                     xvv = xvv + vv * parcels%position(1, n)
+!                     zvv = zvv + vv * parcels%position(2, n)
 !
-!                     x2vv = x2vv + vv * parcels%position(n, 1) ** 2
-!                     z2vv = z2vv + vv * parcels%position(n, 2) ** 2
-!                     xzvv = xzvv + vv * parcels%position(n, 2) * parcels%position(n, 2)
+!                     x2vv = x2vv + vv * parcels%position(1, n) ** 2
+!                     z2vv = z2vv + vv * parcels%position(2, n) ** 2
+!                     xzvv = xzvv + vv * parcels%position(2, n) * parcels%position(2, n)
```
Comment (sjboeing) [inline src/3d/parcels/parcel_diagnostics.f90:249]: Same here.

File: src/3d/parcels/parcel_interpl.f90 [hunk 3]
```diff
@@ -385,47 +378,67 @@ end subroutine grid2par_add
         ! @param[out] jj meridional grid points for interpolation
         ! @param[out] kk vertical grid points for interpolation
         ! @param[out] ww interpolation weights
-        subroutine trilinear(pos, ii, jj, kk, ww)
+        pure subroutine trilinear(pos, ii, jj, kk, ww)
```
Comment (sjboeing) [inline src/3d/parcels/parcel_interpl.f90:381]: Just a quick check: this is without the later optimisations.

File: unit-tests/3d/test_gradient_correction_3d.f90 [hunk 4]
```diff
@@ -70,10 +70,10 @@ program test_gradient_correction_3d
     parcels%B(:, :) = zero
 
     ! b11
-    parcels%B(1:n_parcels, 1) = get_abc(parcels%volume(1:n_parcels)) ** f23
+    parcels%B(1, :) = get_abc(parcels%volume) ** f23
```
Comment (sjboeing) [inline unit-tests/3d/test_gradient_correction_3d.f90:73]: This just makes me think again about places in the code where we need to explicitly ensure that we only look at the parcels we actually use!

**Review by sjboeing — COMMENTED** [review]
Hi Matthias, I just left a few comments, some of them are not even related to your changes!

File: src/3d/parcels/parcel_diagnostics.f90 (hunk 1 above)
Comment (matt-frey) [reply src/3d/parcels/parcel_diagnostics.f90:239]: Yes, you are right. This is wrong! I will fix in the 2D version and anyway remove it here.

File: src/3d/parcels/parcel_interpl.f90 (hunk 3 above)
Comment (matt-frey) [reply src/3d/parcels/parcel_interpl.f90:381]: Yes.

File: src/3d/parcels/parcel_diagnostics.f90 (hunk 1 above)
Comment (matt-frey) [reply src/3d/parcels/parcel_diagnostics.f90]: I fixed in https://github.com/matt-frey/epic/pull/265/files

**Pull request merged**